Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+
+namespace FileSystemVisitorLibrary
+{
+    public class FileSystemVisitor
+    {
+        private readonly Func<FileSystemInfo, bool> _filterFunc;
+
+        private readonly IFileSystem _fileSystem;
+
+        private Actions? _action = Actions.Continue;
+
+        public FileSystemVisitor(IFileSystem fileSystem, Func<FileSystemInfo, bool> filterFunc = null)
+        {
+            _fileSystem = fileSystem;
+            _filterFunc = filterFunc;
+        }
+
+        public event Action Started;
+
+        public event Action Complited;
+
+        public event Func<FileSystemInfo, Actions> FileFound;
+
+        public event Func<FileSystemInfo, Actions> DirectoryFound;
+
+        public event Func<FileSystemInfo, Actions> FilteredFileFound;
+
+        public event Func<FileSystemInfo, Actions> FilteredDirectoryFound;
+
+        public IEnumerable<string> Search(string path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(path, "Directory path is null");
+            }
+
+            if (path == string.Empty)
+            {
+                throw new ArgumentException("Directory path is empty", path);
+            }
+
+            if (!_fileSystem.Directory.Exists(path))
+            {
+                throw new ArgumentException("Selected directory is not exist", path);
+            }
+
+            Started?.Invoke();
+
+            foreach (var f in FileSearch(path))
+            {
+                yield return f;
+            }
+
+            foreach (var d in DirectionSearch(path))
+            {
+                yield return d;
+            }
+
+            Complited?.Invoke();
+        }
+
+        private IEnumerable<string> FileSearch(string path)
```

Comment:
I would recommend to work with `FileSystemInfo` class.

```suggestion
        private IEnumerable<string> FileSearch(FileSystemInfo fileInfo)
```

Reply:
fixed